File: src/neetcode150/go/encode_and_decode_strings.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"unicode/utf8"
)

type Solution struct{}

func (s *Solution) Encode(strs []string) string {
	var tokens []string
	for _, str := range strs {
		tokens = append(tokens, fmt.Sprintf("%d#%s", len(str), str))
	}
	return strings.Join(tokens, "")
}

func (s *Solution) Decode(str string) []string {
	var strs []string
	var start, curr int
	for curr < utf8.RuneCountInString(str) {
		for str[curr] != '#' {
			curr++
		}
		size, _ := strconv.Atoi(str[start:curr])
		strs = append(strs, str[curr+1:curr+1+size])
		start = curr + 1 + size
		curr = start
	}
	return strs
}

File: DSA/src/py/encode_and_decode_strings.py
from typing import List
class Solution:
    DELIMITER = "#"
    def encode(self, strs: List[str]) -> str:
        tokens = []
        for s in strs:
            token = f"{len(s)}{self.DELIMITER}{s}"
            tokens.append(token)

        return "".join(tokens)

    def decode(self, s: str) -> List[str]:
        start = curr = 0
        strs = []
        while curr < len(s):
            while s[curr] != self.DELIMITER:
                curr+=1

            size = int(s[start:curr])
            strs.append(s[curr+1: curr+size+1])
            start = curr+size+1
            curr = curr+size+1

        return strs